crates/subspace-farmer/src/bin/subspace-farmer/commands/shared/network.rs

```diff
@@ -120,17 +133,79 @@ where
         allow_non_global_addresses_in_dht: allow_private_ips,
         known_peers_registry,
         request_response_protocols: vec![
-            PieceByIndexRequestHandler::create(move |_, &PieceByIndexRequest { piece_index }| {
+            {
+                let maybe_weak_node = Arc::clone(&maybe_weak_node);
+                let farmer_cache = farmer_cache.clone();
+
+                CachedPieceByIndexRequestHandler::create(move |peer_id, request| {
+                    let CachedPieceByIndexRequest {
+                        piece_index,
+                        mut cached_pieces,
+                    } = request;
+                    debug!(?piece_index, "Cached piece request received");
+
+                    let maybe_weak_node = Arc::clone(&maybe_weak_node);
```

Comment:
Can you comment on the weak node? Why do we need this wrapper?

Reply:
`Weak` is typically used to break reference cycles and prevent leaks.

In this case we have a request handler that is stored in `Behavior` in `NodeRunner`, which is kept alive as long as at least one `Node` is alive. Storing `Node` inside of it will create a cycle and prevent network from ever shutting down, which we don't want.